src/mod_sasl2.erl

```diff
@@ -0,0 +1,442 @@
+
+-spec init_mod_state(not_provided | exml:element()) -> invalid_agent | mod_state().
+init_mod_state(not_provided) ->
+    #{authenticated => false, id => not_provided, software => not_provided, device => not_provided};
+init_mod_state(El) ->
+    MaybeId = exml_query:attr(El, <<"id">>, not_provided),
+    case if_provided_then_is_not_invalid_uuid_v4(MaybeId) of
+        invalid_agent ->
+            invalid_agent;
+        Value ->
+            Software = exml_query:path(El, [{element, <<"software">>}, cdata], not_provided),
+            Device = exml_query:path(El, [{element, <<"device">>}, cdata], not_provided),
+            #{authenticated => false, id => Value, software => Software, device => Device}
+    end.
+
+-spec if_provided_then_is_not_invalid_uuid_v4(not_provided | binary()) ->
+    invalid_agent | uuid:uuid().
+if_provided_then_is_not_invalid_uuid_v4(not_provided) ->
+    not_provided;
+if_provided_then_is_not_invalid_uuid_v4(Binary) ->
+    try
+        Uuid = uuid:string_to_uuid(Binary),
+        true = uuid:is_v4(Uuid),
+        Uuid
+    catch
+        exit:badarg:_ -> invalid_agent;
+        error:badmatch:_ -> invalid_agent
+    end.
+
+-spec feature(mongoose_c2s:data(), [exml:element()]) -> exml:element().
+feature(C2SData, Mechanisms) ->
+    InlineFeatures = mongoose_hooks:sasl2_stream_features(C2SData, []),
+    InlineElem = inlines(InlineFeatures),
+    #xmlel{name = feature_name(),
+           attrs = [?XMLNS_SASL2],
+           children = [InlineElem | Mechanisms]}.
+
+-spec inlines([exml:element()]) -> exml:element().
+inlines(InlineFeatures) ->
+    #xmlel{name = <<"inline">>, children = InlineFeatures}.
+
+-spec feature_name() -> binary().
+feature_name() ->
+    <<"authentication">>.
+
+-spec get_acc_sasl2_state(mongoose_acc:t()) -> [{module(), inline_request()}].
+get_acc_sasl2_state(SaslAcc) ->
+    mongoose_acc:get(?MODULE, SaslAcc).
+
+-spec get_inline_request(mongoose_acc:t(), module()) -> undefined | inline_request().
+get_inline_request(SaslAcc, ModuleRequest) ->
+    mongoose_acc:get(?MODULE, ModuleRequest, undefined, SaslAcc).
+
+-spec put_inline_request(mongoose_acc:t(), module(), exml:element()) -> mongoose_acc:t().
+put_inline_request(SaslAcc, ModuleRequest, XmlRequest) ->
+    Request = #{request => XmlRequest, response => undefined, status => pending},
+    mongoose_acc:set(?MODULE, ModuleRequest, Request, SaslAcc).
+
+-spec update_inline_request(mongoose_acc:t(), module(), exml:element(), status()) -> mongoose_acc:t().
+update_inline_request(SaslAcc, ModuleRequest, XmlResponse, Status) ->
+    case mongoose_acc:get(?MODULE, ModuleRequest, undefined, SaslAcc) of
+        undefined ->
+            SaslAcc;
+        Request ->
+            Request1 = Request#{response := XmlResponse, status := Status},
+            mongoose_acc:set(?MODULE, ModuleRequest, Request1, SaslAcc)
+    end.
+
+%% Here we extract these two values after all modifications by the inline requests
```

Comment:
"two values"?

Reply:
I meant state and data but yeah actually just "values" is more correct.